### DISCUSSION TITLE
NetSparkleUpdater 3.0.0

### DISCUSSION THREAD

**Discussion by Deadpikle** [NetSparkleUpdater/NetSparkle#632] [General]
# NetSparkleUpdater 3.0.0

## Summary of 3.0.0 Changes

* Updated .NET and .NET Framework compatibility (Bump minimum requirement to .NET 4.6.2, Add .NET 9)
* Built-in JSON app cast reading/writing instead of just XML
* Built-in ability to use different channels for your apps (e.g. beta, alpha, preview)
* Semver compatibility
* Reworked app cast serializing/deserializing API
* New assembly accessor
* Trimming/AOT compatibility
* Nullability annotations
* Smaller file size due to fewer dependencies
* Simplified threading model
* Simplified custom UI interface/usage
* More samples for end users
* Updated README and documentation
* A bunch of bug fixes and adjustments everywhere

## Full change list

For a list of all changes since 2.3.0, including breaking changes, please see: https://github.com/NetSparkleUpdater/NetSparkle/blob/develop/UPGRADING.md

## Note on app cast generator features

To fully utilize some newer features from the app cast generator, e.g. semver, a newer version of the app cast generator is required. Run `dotnet tool update NetSparkleUpdater.Tools.AppCastGenerator -g` to update to the latest version of the app cast generator.

<hr /><em>This discussion was created from the release <a href='https://github.com/NetSparkleUpdater/NetSparkle/releases/tag/3.0.0'>NetSparkleUpdater 3.0.0</a>.</em>

**Comment by balaji222528** [comment]
Thanks for the update. I see you had replaced XMLAppCast File. Can you guide us which will be the one we should be calling instead of XMlAppCast ?Moreover we are still using AppCast XML and customizing that for our needs.Having said that, this new upgrade will affect those ?

**Reply by Deadpikle** [reply]
For your own custom app cast parsing/serializing, please implement your own `IAppCastGenerator` (see `XMLAppCastGenerator` and `JsonAppCastGenerator` for examples) and set the `SparkleUpdater.AppCastGenerator` property to your subclass.

Basically, the app cast handling was separated from the parsing/serialization, which is why there is no more `XMLAppCast` class.

**Comment by balaji222528** [comment]
Also I find these Methods are removed.  public virtual bool downloadParse . Kindly suggest as to  which suitable method we can use instead of this. I'm guessing is it DownloadAppCadt() method in the AppCasthelper ?

**Reply by Deadpikle** [reply]
It depends on what you were trying to do. In 2.x, that method did both downloading and parsing, but now those are two separate functions in different places with different interfaces.

* For downloading, implement `IAppCastDataDownloader` class and set `SparkleUpdater.AppCastDataDownloader`
* For parsing/serializing, implement `IAppCastGenerator` and set `SparkleUpdater.AppCastGenerator`.
* For full control (may or may not be needed by your use case), override `AppCastHelper.DownloadAppCast`, as you noted. You will probably also still need to set `SparkleUpdater.AppCastGenerator` if you need to do custom work on the app cast formatting.

**Reply by Deadpikle** [reply]
Also, for filtering of your own items, implement your own `IAppCastFilter` and set `SparkleUpdater.AppCastHelper.AppCastFilter`.
